site/design-notes/state-of-valhalla/03-vm-model.md

```diff
@@ -613,12 +622,28 @@ and that `I` is really (somehow) an exotic (`Q`-like) descriptor for
 `int.val`.  The `L` descriptor returns `int.ref.class`, while the
 exotic descriptor returns `int.val.class`.
 
+[ Which mirror defines the fields for apis like `getFields()`?]
+
 In particular, `Class::forName` and `Object::getClass` will always return a
 `ref` projection.  If you want the mirror that "works like `int.class`", you
 have to look somewhere special, as with `int` today, or ask the primary mirror
 for its corresponding secondary mirror.  There will be an API point on `Class`
 for obtaining the `val` mirror from the `ref` mirror, and vice versa.
 
+### JVMTI & redefinition of mirrors
+JVMTI, and the `java.lang.instrument` package provide powerful tools for
+modifying classes.  The native `ClassFileLoadHook` allows modifying classes
+before they are loaded and may be used to transform classes in surprising ways
+such as adding methods or fields, and can be used in ways that javac would not,
+violating invariants such as empty constructors.  Both tools can be used to
+modify the bodies of methods after they have been loaded (assuming the classes
+are not marked as `unmodifiable`).
+
+* How do we treat the two mirrors generated from a single primitive classfile?
+* Is the `ref` version marked as non-modifable (treating the `val` one as the `true`
+class)?
+* Does the CFLH get called once (for the classfile) or twice (for `val` & `ref`)?
```

Comment:
From today's call, there is a single set of metadata so 1 CFLH will be called.  Redefining either mirror is OK and affects both.